lib/Infer/AbstractInterpreter.cpp

```diff
@@ -45,11 +45,82 @@ namespace {
     return Max;
   }
 
+ ConstantRange KBToCR(const KnownBits &KB) {
+    ConstantRange CR1(getSMin(KB), getSMax(KB));
+    ConstantRange CR2(getSMin(KB), getSMax(KB));
+    return CR1.getSetSize().ult(CR2.getSetSize()) ? CR1 : CR2;
+  }
+
+  KnownBits CRToKB(const ConstantRange &CR) {
+    KnownBits Res(CR.getBitWidth());
+    if (CR.isFullSet())
+      return Res;
+    if (CR.isEmptySet()) {
+      Res.setAllZero();
+      return Res;
+    }
+    if (CR.isWrappedSet())
+      return Res; // TODO look for opportunities in wrapped case
+    const APInt &L = CR.getLower();
+    const APInt &U = CR.getUpper();
+    for (unsigned I = 0; I < L.getBitWidth(); ++I) {
+      // initial guess is bit from lower bound
+      bool V = L[I];
+      APInt Boundary = L;
+      // gotta be multiple faster ways to do this...
+      for (unsigned J = 0; J < I; ++J)
+        Boundary.setBit(J);
+      if (U.ugt(Boundary + 1))
+        continue;
+      if (V)
+	Res.One.setBit(I);
+      else
+	Res.Zero.setBit(I);
+    }
+    return Res;
+  }
 
+  unsigned numKnown(const KnownBits &KB) {
+    return KB.Zero.countPopulation() + KB.One.countPopulation();
+  }
+
+  KnownBits intersect(const KnownBits &KB1, const KnownBits &KB2) {
```

Comment:
wait, this looks like a union operation, not an intersection. please double-check this code @pranavk

Reply:
I mean that I would expect "intersect" to return the abstraction of the intersection of the concretization sets, but this function returns the abstraction of the union of the concretization sets

Reply:
yeah, i just took this code as it is from what you had in your branch[1]. But indeed, this says intersection but is actually a union. Anyhow, I will fix it as you confirmed that this is incorrect.

>  I mean that I would expect "intersect" to return the abstraction of the intersection of the concretization sets, but this function returns the abstraction of the union of the concretization sets

It takes two abstract KnownBits as parameters, not concretization sets. So, it just returns KB abstraction from two KB abstractions.

[1] https://github.com/regehr/souper/commit/cc95254608aeef5c8038b602d0a8a77ba3394c2f#diff-e2d75e847ce8ee98960879bf86c6ea33R95

Reply:
so how could any of this code be working at all if this basic operation was wrong? please figure out what is going on here, I don't want to merge code that isn't working at all.

in other words, if the testing code had a very basic mistake and it wasn't caught via the testing code, then this testing code is bogus and needs to be fixed. how can we be confident in this code now? this is all making me much less confident in it.

Reply:
"It takes two abstract KnownBits as parameters, not concretization sets. So, it just returns KB abstraction from two KB abstractions."

I'm talking about the specification of the code, not the implementation

Reply:
> so how could any of this code be working at all if this basic operation was wrong? please figure out what is going on here, I don't want to merge code that isn't working at all.

wait, this function `intersect` was used by `improveKBCR` function, part of the changes I picked from your branch, *only*. And testing infrastructure never said that `improveKBCR` is sound.

> in other words, if the testing code had a very basic mistake and it wasn't caught via the testing code, then this testing code is bogus and needs to be fixed. how can we be confident in this code now? this is all making me much less confident in it.

since this bug was only in the function used by `improveKBCR` function, and testing code never promised that `improveKBCR` is correct, I don't think we can conclude that testing code is bogus (yet :).